File: 문제풀이/chap5-5/prob5-5-2.cpp
#include <iostream>
using namespace std;

int N;
double A[509], B[509], C[509];

bool check(double x, double y) {
	for (int i = 1; i <= N; i++) {
		if (A[i] * x + B[i] * y > C[i]) return false;
	}
	return true;
}

int main() {
	// 입력
	cin >> N;
	for (int i = 1; i <= N; i++) cin >> A[i] >> B[i] >> C[i];
	
	// 교점 전체 탐색
	double Answer = 0.0;
	for (int i = 1; i <= N; i++) {
		for (int j = i + 1; j <= N; j++) {
			// 교점을 갖지 않는 경우
			if (A[i] * B[j] == A[j] * B[i]) continue;
			
			// i번째 직선(조건식의 경계)와 j번째 직선(조건식의 경계) 사이의 교점을 구함
			double px = (C[i] * B[j] - C[j] * B[i]) / (A[i] * B[j] - A[j] * B[i]);
			double py = (C[i] * A[j] - C[j] * A[i]) / (B[i] * A[j] - B[j] * A[i]);
			bool ret = check(px - 0.00000001, py - 0.00000001);
			if (ret == true) {
				Answer = max(Answer, px + py);
			}
		}
	}
	
	// 답 출력
	printf("%.12lf\n", Answer);
	return 0;
}

File: 문제풀이/chap5-5/prob5-5-2.py
# 입력
N = int(input())
A = [0.0 for i in range(N)]
B = [0.0 for i in range(N)]
C = [0.0 for i in range(N)]
for i in range(N):
	A[i], B[i], C[i] = map(float, input().split())

# 교점 전체 탐색
Answer = 0.0
for i in range(0, N):
	for j in range(i + 1, N):
		# 교점을 갖지 않는 경우
		if A[i] * B[j] == A[j] * B[i]:
			continue
		
		# i번째 직선(조건식의 경계)와 j번째 직선(조건식의 경계) 사이의 교점을 구함
		px = (C[i] * B[j] - C[j] * B[i]) / (A[i] * B[j] - A[j] * B[i]);
		py = (C[i] * A[j] - C[j] * A[i]) / (B[i] * A[j] - B[j] * A[i]);
		
		# 좌표 (px, py)가 N개의 조건을 만족하는지 확인
		# C++ 코드에서 check 함수에 해당
		ret = True
		for k in range(N):
			if A[k] * px + B[k] * py > C[k]:
				ret = False
		
		if ret == True:
			Answer = max(Answer, px + py)

# 출력
print("{:.12f}".format(Answer))